woocommerce/google-listings-and-ads
# Avoid the test-data plugin occasionally missing in the E2E test environment

### Changes proposed in this Pull Request:

Closes #2175

💡 The issue probably only happens on **MacOS**. Usually I can reproduce it after running the E2E tests 1-3 rounds.

Although I'm not fully confident this PR can fix it, after some time and validation, it seems to be working so far. The solution was mentioned by @mikkamp in https://github.com/woocommerce/google-listings-and-ads/pull/2080#issuecomment-1698679989.

This PR:

- Fix the indentation for the `.wp-env.json` file first as it was mixed with spaces and tabs.
- Make the `test-data.php` plugin have a dedicated directory and change to use [the `plugins` directive](https://github.com/WordPress/gutenberg/tree/%40wordpress/env%409.4.0/packages/env#wp-envjson) to set it up.

### Detailed test instructions:

💡 Considering the issue seems to only happen on MacOS, it would be more appropriate to have someone who uses the same OS or can reproduce the issue to review and test it.

1. Running `npm run wp-env destroy` should not be necessary but it would be ideal to have a clean start to test it.
2. Run `npm run wp-env:up` to see if the E2E test env can be started without errors.
3. Log in http://localhost:8889/wp-admin/ and go to the **Installed Plugins** page.
4. Check if the **GLA Test Data** plugin is activated.
5. Run `npm run test:e2e` to see if all E2E tests can pass.
6. Refresh the **Installed Plugins** page to see if the **GLA Test Data** plugin is activated.
7. Repeat steps 5-6 several times to verify if the plugin missing issue no longer occurs.

### Additional details:

#### 📌 Investigation

Although I didn't find the exact cause, the process that led to the issue is roughly as follows.

1. From the docker-compose.yml generated by `wp-env`, it's shown there are nested mounts under the `/var/www/html` directory.
   ```yml
   version: '3.7'
   services:
     # ... omitted
     tests-wordpress:
       # ... omitted
       volumes:
         - '/Users/eason/.wp-env/fa3f9473ab3c866d9e18213e3637074d/tests-WordPress:/var/www/html'
         - '/Users/eason/.wp-env/fa3f9473ab3c866d9e18213e3637074d/tests-WordPress-PHPUnit/tests/phpunit:/wordpress-phpunit'
         - 'tests-user-home:/home/eason'
         - '/Users/eason/wp-local-docker-sites/wp-test/wordpress/wp-content/plugins/google-listings-and-ads/tests/e2e/config/wp-cli.yml:/var/www/html/wp-cli.yml'
         - '/Users/eason/wp-local-docker-sites/wp-test/wordpress/wp-content/plugins/google-listings-and-ads/tests/e2e/bin/test-data.php:/var/www/html/wp-content/plugins/test-data.php'
         - '/Users/eason/.wp-env/fa3f9473ab3c866d9e18213e3637074d/master:/var/www/html/wp-content/plugins/master'
         - '/Users/eason/wp-local-docker-sites/wp-test/wordpress/wp-content/plugins/google-listings-and-ads:/var/www/html/wp-content/plugins/google-listings-and-ads'
   
   ```
2. In the container, the `test-data.php` is mounted in the beginning.
   ```bash
   root@969c943abe08:/var/www/html# mount | grep www
   grpcfuse on /var/www/html type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
   grpcfuse on /var/www/html/wp-cli.yml type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
   grpcfuse on /var/www/html/wp-content/plugins/test-data.php type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
   grpcfuse on /var/www/html/wp-content/plugins/google-listings-and-ads type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
   grpcfuse on /var/www/html/wp-content/plugins/master type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
   ```
   In the container, its file size is 3.0K.
   ```bash
   root@969c943abe08:/var/www/html# ls -lh /var/www/html/wp-content/plugins/test-data.php
   -rw-r--r-- 1 eason dialout 3.0K Feb 29 07:02 /var/www/html/wp-content/plugins/test-data.php
   ```
3. Due to the nested mounts, a `test-data.php` file is created back to the host side with **0 file size**.
   <img width="1637" alt="1" src="https://github.com/woocommerce/google-listings-and-ads/assets/17420811/2de16216-605f-4498-a4ea-16405c297a20">
4. After a few runs of E2E test, somehow the `test-data.php` is unmounted.
   ```bash
   root@969c943abe08:/var/www/html# mount | grep www
   grpcfuse on /var/www/html type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
   grpcfuse on /var/www/html/wp-content/plugins/google-listings-and-ads type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
   grpcfuse on /var/www/html/wp-content/plugins/master type fuse.grpcfuse (rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other,max_read=1048576)
   ```
   In the container, its file size is 0 now. I think the empty content file is synced from the host.
   (See the screenshot in step 3)
   ```bash
   root@969c943abe08:/var/www/html# ls -lh /var/www/html/wp-content/plugins/test-data.php
   -rwxr-xr-x 1 eason dialout 0 Feb 29 07:06 /var/www/html/wp-content/plugins/test-data.php
   ```
5. Now the issue occurs.
   - The Installed Plugins page
      ![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/05c1eefe-a729-4651-a686-083d7167edb2)
   - E2E test always failed
      ![2](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/e1c95bf4-3b0b-4eb8-9fea-d9fc9348fb06)

I couldn't find the same issue report, but I suspect the reason for this is similar to these past issues:

- https://github.com/docker/compose/issues/6975
- https://github.com/docker/for-mac/issues/3479
- https://github.com/docker/for-mac/issues/3976
- https://github.com/docker/for-mac/issues/2777

#### 📌 About the `wp-cli.yml` file

Please note that the issue also happens on the `wp-cli.yml` file. If I understood it right, it's a one-time setup during initialization and won't break when the `wp-cli.yml` file becomes unavailable. Therefore, this PR doesn't adjust it.

https://github.com/woocommerce/google-listings-and-ads/blob/e6c03c151b7b4eff133579dffbd3e58164bd97d6/tests/e2e/bin/test-env-setup.sh#L9-L10

### Changelog entry

> Dev - Avoid the test-data plugin occasionally missing in the E2E test environment
